packages/plugin-dkg/src/actions/dkgInsert.ts

```diff
@@ -0,0 +1,223 @@
+        (msg) => msg.roomId === relatedRoomId
+    );
+
+    const concatenatedText = relatedMessages
+        .map((msg) => msg.content.text)
+        .join(" ");
+
+    return concatenatedText;
+}
+
+const DkgClient = new DKG({
+    environment: process.env.DKG_ENVIRONMENT,
+    endpoint: process.env.DKG_HOSTNAME,
+    port: process.env.DKG_PORT,
+    blockchain: {
+        name: process.env.DKG_BLOCKCHAIN_NAME,
+        publicKey: process.env.DKG_PUBLIC_KEY,
+        privateKey: process.env.DKG_PRIVATE_KEY,
+    },
+    maxNumberOfRetries: 300,
+    frequency: 2,
+    contentType: "all",
+    nodeApiVersion: "/v1",
+});
+
+async function constructKnowledgeAsset(
+    runtime: IAgentRuntime,
+    userQuery: string,
+    additionalContext: string,
+    state?: State
+) {
+    const context = `
+  You are tasked with creating a structured memory JSON-LD object for an AI agent. The memory represents the interaction captured via social media. Your goal is to extract all relevant information from the provided user query and additionalContext which contains previous user queries (only if relevant for the current user query) to populate the JSON-LD memory template provided below.
+
+  ** Template **
+  The memory should follow this JSON-LD structure:
+  ${JSON.stringify(dkgMemoryTemplate)}
+
+  ** Instructions **
+  1. Extract the main idea of the user query and use it to create a concise and descriptive title for the memory. This should go in the "headline" field.
+  2. Store the original post in "articleBody".
+  3. Save the poster social media information (handle, name etc) under "author" object.
+  4. For the "about" field:
+     - Identify the key topics or entities mentioned in the user query and add them as Thing objects.
+     - Use concise, descriptive names for these topics.
+     - Where possible, create an @id identifier for these entities, using either a provided URL, or a well known URL for that entity. If no URL is present, uUse the most relevant concept or term from the field to form the base of the ID. @id fields must be valid uuids or URLs
+  5. For the "keywords" field:
+     - Extract relevant terms or concepts from the user query and list them as keywords.
+     - Ensure the keywords capture the essence of the interaction, focusing on technical terms or significant ideas.
+  6. Ensure all fields align with the schema.org ontology and accurately represent the interaction.
+  7. Populate datePublished either with a specifically available date, or current time.
+
+  ** Input **
+  User Query: ${userQuery}
+  Previous conversation that you had (can be with another user): ${additionalContext}
+
+  ** Output **
+  Generate the memory in the exact JSON-LD format provided above, fully populated based on the input query.
+  Make sure to only output the JSON-LD object. DO NOT OUTPUT ANYTHING ELSE, DONT ADD ANY COMMENTS OR REMARKS, JUST THE JSON LD CONTENT WRAPPED IN { }.
+  `;
+    const content = await generateText({
```

Comment:
We can leverage generateObject to get a type-safe object see packages/plugin-coinbase/src/plugins/trade.ts as an example. You can then remove the parsing logic